core/src/main/java/com/google/googlejavaformat/java/Main.java

```diff
@@ -142,28 +143,45 @@ private int formatFiles(CommandLineOptions parameters, JavaFormatterOptions opti
       } catch (ExecutionException e) {
         if (e.getCause() instanceof FormatterException) {
           for (FormatterDiagnostic diagnostic : ((FormatterException) e.getCause()).diagnostics()) {
-            errWriter.println(result.getKey() + ":" + diagnostic.toString());
+            errWriter.println(path + ":" + diagnostic.toString());
           }
         } else {
-          errWriter.println(result.getKey() + ": error: " + e.getCause().getMessage());
+          errWriter.println(path + ": error: " + e.getCause().getMessage());
           e.getCause().printStackTrace(errWriter);
         }
         allOk = false;
         continue;
       }
+      boolean idempotent = formatted.equals(inputs.get(path)); // not changed
+      if (parameters.isValidationMode()) {
+        if (idempotent) {
+          continue; // no formatting needed, validate next
+        }
+        errWriter.println(path + ": needs formatting");
+        if (parameters.isSetExitIfChanged()) {
+          allOk = false;
+        }
+        continue;
+      }
       if (parameters.inPlace()) {
-        if (formatted.equals(inputs.get(result.getKey()))) {
+        if (idempotent) {
           continue; // preserve original file
         }
         try {
-          Files.write(result.getKey(), formatted.getBytes(UTF_8));
+          Files.write(path, formatted.getBytes(UTF_8));
         } catch (IOException e) {
-          errWriter.println(result.getKey() + ": could not write file: " + e.getMessage());
+          errWriter.println(path + ": could not write file: " + e.getMessage());
           allOk = false;
           continue;
         }
-      } else {
-        outWriter.write(formatted);
+        if (parameters.isSetExitIfChanged()) {
+          allOk = false;
+        }
```

Comment:
Can you move this `if` to immediately below the declaration of idempotent/changed to avoid some duplication?

Reply:
Not so easy. The default "stdout" is implemented as a fall-through or "catch-all-else". Will try.

Reply:
Sorry, what am I missing?

```
index 7b7c873..b331a78 100644
--- a/core/src/main/java/com/google/googlejavaformat/java/Main.java
+++ b/core/src/main/java/com/google/googlejavaformat/java/Main.java
@@ -153,14 +153,14 @@ public final class Main {
         continue;
       }
       boolean idempotent = formatted.equals(inputs.get(path)); // not changed
+      if (!idempotent && parameters.isSetExitIfChanged()) {
+        allOk = false;
+      }
       if (parameters.isValidationMode()) {
         if (idempotent) {
           continue; // no formatting needed, validate next
         }
         errWriter.println(path + ": needs formatting");
-        if (parameters.isSetExitIfChanged()) {
-          allOk = false;
-        }
         continue;
       }
       if (parameters.inPlace()) {
@@ -174,15 +174,9 @@ public final class Main {
           allOk = false;
           continue;
         }
-        if (parameters.isSetExitIfChanged()) {
-          allOk = false;
-        }
         continue;
       }
       outWriter.write(formatted);
-      if (parameters.isSetExitIfChanged() && !idempotent) {
-        allOk = false;
-      }
     }
     return allOk ? 0 : 1;
   }
```

Reply:
I'll push a cleaner solution. It needs logical "is-not-running-in-special-mode" helper in `CommandLineOptions`:

```java
  /** Returns true if no special run-mode (like in-place or dry-run) is selected. */
  boolean stdout() {
    return !(inPlace || dryRun);
  }
```

Reply:
Making the combination a usage error sounds fine.

I'm still not sure I understand what the help is needed for?

Reply:
Maybe I'm assuming s/t wrong here. 

Should the stdout-case **always** print out the potentially formatted text?

Reply:
`google-java-format T.java` always prints formatted text to stdout, regardless of whether the input was formatted. `google-java-format -n T.java` prints `T.java` to stdout if the input was not already formatted.

Reply:
The current iteration behaves like this. I don't see an easy way to re-use a boolean state at the moment.